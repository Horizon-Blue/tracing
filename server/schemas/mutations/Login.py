from graphene import Mutation, String
from graphql import GraphQLError
import jwt
import datetime
from secrets import TOKEN_SECRET
from ..objectTypes import User
import os


def get_tomorrow():
    """
    Helper function to get tomorrow
    :return: datetime represents tomorrow
    """
    if os.environ.get("__TEST__"):
        return datetime.datetime(1901, 12, 21)
    return datetime.datetime.now() + datetime.timedelta(days=1)


def generate_token(user):
    """
    Helper function to generate jwt token
    :param user:
    :return:
    """
    return jwt.encode({
        'sub': user.id,
        'exp': get_tomorrow(),
    },
        TOKEN_SECRET, algorithm='HS256').decode()


class Login(Mutation):
    """
    Given the name and password, generate a token for the user
    """

    class Arguments:
        name = String(required=True)
        password = String(required=True)

    token = String()

    @staticmethod
    def mutate(root, info, name, password):
        """
        Generate the token for the specific user, if exist
        :param root: the root information
        :param info: context object
        :param name: name of user
        :param password: password of user
        :return: token string, if found, or raise error
        """
        user = User.get_query(info).filter_by(name=name).first()
        if user and user.password == password:
            token = generate_token(user)
            return Login(token=token)
        else:
            raise GraphQLError("Incorrect credentials")
