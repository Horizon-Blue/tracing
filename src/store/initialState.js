export default {
  locale: 'en',
  token: null,
};
