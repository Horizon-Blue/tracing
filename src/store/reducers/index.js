import { combineReducers } from 'redux';
import locale from './locale';
import token from './token';

export default combineReducers({ locale, token });
