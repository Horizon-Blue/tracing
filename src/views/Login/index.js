import React, { PureComponent } from 'react';
import { Container, Form } from 'semantic-ui-react';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ErrorMessage from '../ErrorMessage';
import Translated from '../Translated';
import actions from '../../actions';
import './index.less';

class Login extends PureComponent {
  static propTypes = {
    mutate: PropTypes.func.isRequired,
    setToken: PropTypes.func.isRequired,
    token: PropTypes.string,
    history: PropTypes.object.isRequired,
  };

  state = {
    name: '',
    password: '',
  };

  /**
   * Update the current value in the input field to the state
   *
   * @param      {event}  the event being triggered
   */
  handleChange = ({ target: { name, value } }) =>
    this.setState({ [name]: value });

  handleLogin = () => {
    const { name, password } = this.state;
    const { setToken } = this.props;

    this.props
      .mutate({ variables: { name, password } })
      .then(({ data: { login: { token } } }) => setToken(token))
      .catch(error => this.setState({ error }));
  };

  render() {
    if (this.props.token) return <Redirect to="/" />;

    const { name, password, error } = this.state;

    return (
      <Container text as="main" textAlign="center">
        <Form inverted className="login-form">
          <Form.Group widths="equal">
            <Form.Field>
              <Translated as="label" id="name" />
              <input value={name} name="name" onChange={this.handleChange} />
            </Form.Field>
            <Form.Field>
              <Translated as="label" id="password" />
              <input
                type="password"
                name="password"
                value={password}
                onChange={this.handleChange}
              />
            </Form.Field>
          </Form.Group>
          <Translated
            as={Form.Button}
            inverted
            basic
            color="blue"
            type="submit"
            id="login"
            onClick={this.handleLogin}
          />
        </Form>
        {!!error && <ErrorMessage value={error} />}
      </Container>
    );
  }
}

const loginMutation = gql`
  mutation($name: String!, $password: String!) {
    login(name: $name, password: $password) {
      token
    }
  }
`;

const mapStateToProps = state => ({ token: state.get('token') });

const mapDispatchToProps = dispatch => ({
  setToken: token =>
    dispatch({
      type: actions.SET_TOKEN,
      token,
    }),
});

export default connect(mapStateToProps, mapDispatchToProps)(
  graphql(loginMutation)(Login)
);
