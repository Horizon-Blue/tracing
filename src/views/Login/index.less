main .login-form {
  margin-top: 3vh;
}

.equal.width.fields .field input {
  background: none;
  border-width: 2px;
  border-color: rgba(190, 220, 255, 0.8);
  color: white;
}
