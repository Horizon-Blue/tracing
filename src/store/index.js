import { createStore } from 'redux';
import initialState from './initialState';
import rootReducer from './reducers';

const store = createStore(
  rootReducer,
  initialState,
  /* Enable Redux Devtools Chrome Extension */
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);

export default store;
