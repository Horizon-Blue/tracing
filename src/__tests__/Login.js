import React from 'react';
import Login from '../views/Login';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { ApolloProvider } from 'react-apollo';
import { Provider } from 'react-redux';
import store from '../store';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <ApolloProvider client={{}}>
        <Provider store={store}>
          <BrowserRouter>
            <Login />
          </BrowserRouter>
        </Provider>
      </ApolloProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
