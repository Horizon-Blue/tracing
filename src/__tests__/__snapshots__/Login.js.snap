// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  className="ui text center aligned container"
>
  <form
    action={undefined}
    className="ui inverted form login-form"
    onSubmit={[Function]}
  >
    <div
      className="equal width fields"
    >
      <div
        className="field"
      >
        <label
          dispatch={undefined}
        >
          Name
        </label>
        <input
          name="name"
          onChange={[Function]}
          value=""
        />
      </div>
      <div
        className="field"
      >
        <label
          dispatch={undefined}
        >
          Password
        </label>
        <input
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <button
        className="ui blue basic inverted button"
        disabled={undefined}
        dispatch={undefined}
        onClick={[Function]}
        required={undefined}
        role="button"
        tabIndex={undefined}
        type="submit"
      >
        Login
      </button>
    </div>
  </form>
</main>
`;
